plugins/CoreHome/vue/src/EntityDuplicator/EntityDuplicatorAdapter.ts

```diff
@@ -0,0 +1,164 @@
+   * needed for validation.
+   */
+  validateFormFields(
+    formValues: Record<string, unknown>,
+  ): ValidationResult | Promise<ValidationResult>;
+
+  /**
+   * Prepares the API parameters for the duplication request. Optionally return promise in case API
+   * request is needed for to prepare all parameters.
+   */
+  prepareApiParams(formValues: Record<string, unknown>): QueryParameters | Promise<QueryParameters>;
+
+  /**
+   * Submits the duplication request to the server and returns a promise for the modal to wait for
+   */
+  submitRequest(params: QueryParameters): Promise<DuplicateRequestResponse>;
+
+  /**
+   * Optional: Called after successful duplication
+   */
+  onSuccess?(response: DuplicateRequestResponse): void;
+
+  /**
+   * Optional: Called after failed duplication
+   */
+  onFailure?(error: DuplicateRequestResponse | Error): void;
+
+  /**
+   * Optional: Called before showing the modal and can be used to show the modal as loading if
+   * there are any async operations needed to finish presenting the modal.
+   */
+  beforeShowModal?(): void | Promise<void>;
+}
+
+export class BaseDuplicatorAdapter implements EntityDuplicatorAdapter {
+  module: string;
+
+  method: string;
+
+  format: string;
+
+  requiredFields: string[];
+
+  constructor(properties: EntityDuplicatorAdapterProperties) {
+    this.module = properties.module || 'API';
+    this.method = properties.method;
+    this.format = properties.format || 'json';
+    this.requiredFields = properties.requiredFields || ['idSite', 'idDestinationSites'];
+  }
+
+  async validateFormFields(
+    formValues: Record<string, unknown>,
+  ): Promise<ValidationResult> {
+    const errorMessages: string[] = [];
+
+    this.requiredFields.forEach((fieldName) => {
+      if (!(fieldName in formValues) || !formValues[fieldName]) {
+        errorMessages.push(translate('General_Required', fieldName));
+      }
+    });
+
+    return new Promise((resolve) => resolve({
+      errorMessages,
+      isValid: errorMessages.length === 0,
+    } as ValidationResult));
+  }
+
+  prepareApiParams(formValues: Record<string, unknown>): QueryParameters {
+    return {
+      idSite: Matomo.idSite || MatomoUrl.parsed.value.idSite,
+      idDestinationSites: [formValues.idDestinationSite as number|string],
+      ...formValues,
+    };
+  }
+
+  async submitRequest(params: QueryParameters): Promise<DuplicateRequestResponse> {
+    // Override the defaults if provided in the params and then remove them from the params
+    this.module = params.module || this.module;
+    this.method = params.method || this.method;
+    this.format = params.format || this.format;
+    const postParams = params as Omit<QueryParameters, 'module' | 'method' | 'format'>;
+
+    if (!this.method || this.method.length < 1) {
+      throw new Error('The POST method cannot be empty!');
+    }
+
+    const ajax = new AjaxHelper();
```

Comment:
```suggestion
    const ajax = new AjaxHelper();
    ajax.useCallbackInCaseOfError();
    ajax.setErrorCallback(null);
```

A fatal error on the PHP side would currently be displayed both inside the modal, and as a regular error toast.

Mimicking the `createErrorNotification` option of the static helper methods will display it inside the modal as the generic error message.

If a backend error should close the modal and display as a toast, I think that could work by checking the `error` in the `catch` callback is an XHR response with a non-200 code. And then close modal + display error toast 🤔

Reply:
Thank you @mneudert . As far as I can tell, that isn't necessary. When I test, the error is only displayed in the modal:
<img width="1638" height="781" alt="image" src="https://github.com/user-attachments/assets/75170418-fa1a-49f4-9a7c-2193ba83f881" />

Maybe it's because I use a catch when processing the promise?

Reply:
It is behaving different from my [goal duplication test branch](https://github.com/matomo-org/matomo/tree/copy-goals) because of the way the heatmap management page is rendered. The (sometimes) used ajax error element is not available:

https://github.com/matomo-org/matomo/blob/be0228d25acdd1643a93d341ea42dbc2ead6e159/plugins/Goals/templates/_addEditGoal.twig#L2-L3

If you add `<div id="ajaxError" style="display:none"></div>` to the heatmap management widget, the errors should appear in both places.

Without that element you can force the duplicate display with a syntax error. If you open the page, then break the code (remove a random `;` or something like that), and then try to duplicate something, there should be two "Oops" messages visible:

<img width="2130" height="864" alt="Screenshot from 2025-07-29 18-09-39" src="https://github.com/user-attachments/assets/25cdc971-9ce9-49a4-a8a4-45ced740847a" />